IPython/core/tests/test_async_helpers.py

```diff
@@ -14,298 +14,297 @@
-            if True:
-                while True:
+    def test_top_level_return_error(self):
+        tl_err_test_cases = self._get_top_level_cases()
+        tl_err_test_cases.extend(self._get_ry_syntax_errors())
+
+        vals = ('return', 'yield', 'yield from (_ for _ in range(3))',
+                dedent('''
+                    def f():
+                        pass
+                    return
+                    '''),
+                )
+
+        for test_name, test_case in tl_err_test_cases:
+            # This example should work if 'pass' is used as the value
+            with self.subTest((test_name, 'pass')):
+                iprc(test_case.format(val='pass'))
+
+            # It should fail with all the values
+            for val in vals:
+                with self.subTest((test_name, val)):
+                    msg = "Syntax error not raised for %s, %s" % (test_name, val)
+                    with self.assertRaises(SyntaxError, msg=msg):
+                        iprc(test_case.format(val=val))
+
+    def test_in_func_no_error(self):
+        # Test that the implementation of top-level return/yield
+        # detection isn't *too* aggressive, and works inside a function
+        func_contexts = []
+
+        func_contexts.append(('func', False, dedent("""
+        def f():""")))
+
+        func_contexts.append(('method', False, dedent("""
+        class MyClass:
+            def __init__(self):
+        """)))
+
+        func_contexts.append(('async-func', True,  dedent("""
+        async def f():""")))
+
+        func_contexts.append(('async-method', True,  dedent("""
+        class MyClass:
+            async def f(self):""")))
+
+        func_contexts.append(('closure', False, dedent("""
+        def f():
+            def g():
+        """)))
+
+        def nest_case(context, case):
+            # Detect indentation
+            lines = context.strip().splitlines()
+            prefix_len = 0
+            for c in lines[-1]:
+                if c != ' ':
                     break
-                    for x in range(3):
-                        if True:
-                            while True:
-                                for x in range(3):
-                                    {val}
-            """)))
+                prefix_len += 1
 
-            return test_cases
+            indented_case = indent(case, ' ' * (prefix_len + 4))
+            return context + '\n' + indented_case
 
-        def _get_ry_syntax_errors(self):
-            # This is a mix of tests that should be a syntax error if
-            # return or yield whether or not they are in a function
+        # Gather and run the tests
 
-            test_cases = []
+        # yield is allowed in async functions, starting in Python 3.6,
+        # and yield from is not allowed in any version
+        vals = ('return', 'yield', 'yield from (_ for _ in range(3))')
+        async_safe = (True,
+                        True,
+                        False)
```

Comment:
Now that `async_safe` is constant; once this is merged; there is probably some boolean logic that can be simplified below. 

We can leave that for another new contributor.